### PR TITLE
[CIS-806] Brand new `MessageContentView`

### PR DESCRIPTION
**Pull request by evsaev**
**This PR:**
- introduces extendable `MessageContentView`
- introduces `MessageLayoutOptions` which determines the layout of `MessageContentView`
- introduces simplified UI components displaying the message and reaction bubble + thread arrow
- adds some CoreGraphics helpers

### CODE REVIEW

**Comment by evsaev** [comment]
🤔 not sure about the following things:
- do we need to update `CHANGELOG.md`. The new content view is just introduced but not used ATM
- if the current naming is OK (there's no `Chat` prefix) and should we deprecate the old message content view?

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    /// Gets called when thread reply button is tapped.
+    /// - Parameter indexPath: The index path of the cell displaying the content view. Equals to `nil` when
+    /// the content view is displayed outside the collection/table view.
+    func messageContentViewDidTapOnThread(_ indexPath: IndexPath?)
+
+    /// Gets called when quoted message view is tapped.
+    /// - Parameter indexPath: The index path of the cell displaying the content view. Equals to `nil` when
+    /// the content view is displayed outside the collection/table view.
+    func messageContentViewDidTapOnQuotedMessage(_ indexPath: IndexPath?)
+}
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift:27]: Following up on you comments about the index path, should we use `messageId` here, instead?

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+
+            mainContainer.addArrangedSubview(avatarView)
+        }
+        
+        if options.contains(.avatarSizePadding) {
+            let spacer = UIView().withoutAutoresizingMaskConstraints
+            spacer.isHidden = true
+            constraintsToActivate += [spacer.widthAnchor.pin(equalToConstant: 32)]
+            mainContainer.addArrangedSubview(spacer)
+        }
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift:183]: Since the avatar size is used in two places, it would be nice to extract it into a variable

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+                reactions: $0.reactions,
+                didTapOnReaction: nil
+            )
+        }
+    }
+
+    override open func tintColorDidChange() {
+        super.tintColorDidChange()
+
+        updateContentIfNeeded()
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift:357]: please add a comment why is this needed

**Review by VojtaStavik — APPROVED** [review]
Looks great 🎉 I have a couple of comments, we can discuss them in Slack.

Let's also talk about the colors in Sources/StreamChatUI/ChatMessageList/RETHINK/__Snapshots__/MessageContentView_Tests/test_appearance.default-light-bubble-avatar-metadata-text-quotedMessage-reactions.png

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageLayoutOptions.swift [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    static let text = Self(rawValue: 1 << 7)
+
+    /// If set the message quoted by the current message will be shown.
+    static let quotedMessage = Self(rawValue: 1 << 8)
+
+    /// If set the message thread replies information will be shown.
+    static let threadInfo = Self(rawValue: 1 << 9)
+
+    /// If set the error indicator will be shown.
+    static let error = Self(rawValue: 1 << 10)
```
[the remark this answers is not in the document]
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/RETHINK/MessageLayoutOptions.swift]: ✅ Done!

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageBubbleView.swift [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+public typealias MessageBubbleView = _MessageBubbleView<NoExtraData>
+
+open class _MessageBubbleView<ExtraData: ExtraDataTypes>: _View, UIConfigProvider {
+    open var roundedCorners: CACornerMask = .all {
+        didSet { updateContentIfNeeded() }
+    }
+
+    open var cornerRadius: CGFloat { 16 }
+
+    open var borderWidth: CGFloat { 1 }
```
[the remark this answers is not in the document]
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/RETHINK/MessageBubbleView.swift]: ✅ Done!

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageMetadataView.swift [hunk 6]
```diff
@@ -8,7 +8,12 @@ import UIKit
 public typealias ChatMessageMetadataView = _ChatMessageMetadataView<NoExtraData>
 
 open class _ChatMessageMetadataView<ExtraData: ExtraDataTypes>: _View, UIConfigProvider {
-    public var message: _ChatMessageGroupPart<ExtraData>? {
+    public struct Content {
+        public let message: _ChatMessage<ExtraData>
+        public let isAuthorNameShown: Bool
+    }
```
[the remark this answers is not in the document]
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageMetadataView.swift]: ✅ Done!

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageMetadataView.swift [hunk 7]
```diff
@@ -17,97 +22,90 @@ open class _ChatMessageMetadataView<ExtraData: ExtraDataTypes>: _View, UIConfigP
...
+        .withBidirectionalLanguagesSupport
+        .withoutAutoresizingMaskConstraints
+
+    public private(set) lazy var eyeImageView = UIImageView()
+        .withoutAutoresizingMaskConstraints
     
+    public private(set) lazy var eyeTextLabel = UILabel()
+        .withAdjustingFontForContentSizeCategory
+        .withBidirectionalLanguagesSupport
+        .withoutAutoresizingMaskConstraints
```
[the remark this answers is not in the document]
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageMetadataView.swift]: ✅ Done and introduced new `onlyVisibleForYouIndicator` layout option

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift (hunk 2 above)
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift]: ✅ Done!

File: Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift (hunk 3 above)
Comment (evsaev) [reply Sources/StreamChatUI/ChatMessageList/RETHINK/MessageContentView.swift]: ✅ Done!
